src/wallet/export/caravan.rs

```diff
@@ -0,0 +1,461 @@
+                    _ => Err(Error::Generic(
+                        "Unsupported sh(wsh()) inner descriptor.".to_string(),
+                    )),
+                },
+                _ => Err(Error::Generic(
+                    "Unsupported sh() inner descriptor.".to_string(),
+                )),
+            },
+            Descriptor::Wsh(sh) => match sh.as_inner() {
+                WshInner::SortedMulti(smv) => {
+                    let (quorum, extended_public_keys) = CaravanExport::parse_sorted_multi(smv);
+                    Ok((CaravanAddressType::P2wsh, quorum, extended_public_keys))
+                }
+                _ => Err(Error::Generic(
+                    "Unsupported wsh() inner descriptor.".to_string(),
+                )),
+            },
+            _ => Err(Error::Generic(
+                "Unsupported top level descriptor.".to_string(),
+            )),
+        }?;
+
+        let network = match network {
+            Network::Bitcoin => CaravanNetwork::Mainnet,
+            _ => CaravanNetwork::Testnet,
+        };
+        let client = CaravanClient { value: client_type };
+        let extended_public_keys: Vec<CaravanExtendedPublicKey> = descriptor_public_keys
+            .iter()
+            .map(|k| match k {
+                DescriptorPublicKey::SinglePub(_) => {
+                    Err(Error::Generic("Unsupported single pub key.".to_string()))
+                }
+                DescriptorPublicKey::XPub(xpub) => {
+                    let mut xfp = None;
+                    let mut bip32_path = None;
+                    if let Some((s_xfp, s_bip32_path)) = xpub.clone().origin {
+                        xfp = Some(s_xfp);
+                        bip32_path = Some(s_bip32_path);
+                    }
+                    Ok(CaravanExtendedPublicKey {
+                        name: xpub.xkey.fingerprint().to_string(),
+                        bip32_path,
+                        xpub: xpub.xkey,
+                        xfp,
+                    })
+                }
+            })
+            .flatten()
+            .collect::<Vec<CaravanExtendedPublicKey>>();
+
+        Ok(Self {
+            name,
+            address_type,
+            network,
+            client,
+            quorum,
+            extended_public_keys,
+            starting_address_index: 0,
+        })
+    }
+}
+
+/// The address types supported by Caravan
+#[derive(Debug, Serialize, Deserialize)]
+pub enum CaravanAddressType {
+    /// P2SH
+    #[serde(rename = "P2SH")]
+    P2sh,
+    /// P2SH-P2WSH
+    #[serde(rename = "P2SH-P2WSH")]
+    P2shP2wsh,
+    /// P2WSH
+    #[serde(rename = "P2WSH")]
+    P2wsh,
+}
+
+/// The networks supported by Caravan
+#[derive(Debug, Serialize, Deserialize)]
+enum CaravanNetwork {
+    #[serde(rename = "mainnet")]
+    Mainnet,
+    #[serde(rename = "testnet")]
+    Testnet,
+}
+
+/// A caravan client
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CaravanClient {
```

Comment:
I guess this could be made into an enum? Is there a list of valid values for the client?

Reply:
I don't know what the values should be but this question is on my list.

Reply:
So I think this is a good argument in favor of the generalized approach detailed in my comment where these details could/should be left up to the consumer of this import/export utility. In Caravan, the client can be public and then we just assume that points to blockstream.info. There's the private one which requires some access info for your node (host, rpcuser, rpcpass), but i don't think bdk needs to care about that.